Introduce winVersion.WinVersion class to record Windows version information, removing winVersion.isWin10 function

josephsl opened the pull request:
Hi,

I think it would be best to target 2021.1:

### Link to issue number:
Closes #11795 
Closes #11837 
Closes #11933 
Replaces #11796 
Replaces #11799 

### Summary of the issue:
Prior to October 2020 Update, Windows 10 feature updates (releases) were integers. Starting with October 2020 Update (build 19042), it is now a string, denoting which half of the year the given release was made public e.g. Version 20H2 denotes second half of 2020. This breaks winVersion.isWin10 function as it expects integers. Also, with the impending release (or advent) of Windows 10 X and in order to better detect Windows 10 in S mode, it became crucial to have a central location to record Windows version information and other properties.

### Description of how this pull request fixes the issue:
Introduce winVersion.WinVersion class to record the following Windows version information:

* Major version (integer)
* Minor version (integer)
* Build (integer)
* Service pack (string)
* Product type (string): one of "workstation", "domain controller", "server".
* Version comparison operators (using functools.total_ordering decorator)
* A repr of Windows release, service pack (if any), major.minor.build tuple, and product type, and private helper methods to obtain these.

Additional functions in winVersion module:

* getWinVer: return WinVersion instance recording the current version of Windows in use.
* getWinVerFromString: returning an instance of WinVersion class based on a string recording major.minor.build e.g. 6.1.7601.

winVersion module will also come with constants representing Windows releases to WinVersion class instances, as well as removing old winVersion.winVersion (sys.getwindowsversion() tuple) and winVersion.winVersionText to reduce developer confusion. With these changes, winVersion.isWin10 function is marked for removal in 2021.1, and affected code paths throughout NVDA will be edited.

### Testing performed:
Tested with:

* Source code (from NVDA itself and via Python interpreter)
* Unit tests

### Known issues with pull request:
If this PR is accepted, various parts of NVDA and add-ons calling winVersion.isWin10 and winVersion.* must be modified to use constants for comparisons. Also, add-ons using old winVersion.winVersion and its properties must be modified to call getWinVer() and friends (notably, this affects Windows 10 App Essentials add-on, for which an issue will be created on that add-ons' repo once this PR is accepted).

### Change log entry:
Changes for developers:

getWinVer function was added to winVersion module to obtain a record of the version of Windows currently in use.
isWin10 function found in winVersion module is removed. To check for a specific Windows 10 release, compare winVersion.getWinVer() with a specific winVersion.WIN10_* constants.

josephsl committed:
WinVersion: update copyright header, add Bill Dengler as author

josephsl committed:
WinVersion: introduce WinVersion class. Re #11837.

Introduce winVersion.WinVersion class to serve as a hub of Windows version information for the system. The class (wrapped inside functools.total_ordering) will record version major, minor, build, and service pack information if present. It will also include a way to check for specific Windows 10 releases i.e. transplant winVersion.isWin10 function to winVersion.WinVersion.isWin10 method (later commit), as well as preparing for other checks such as Windows 10 X, Windows 10 on S mode, server, and others in the future. The class itself is wrapped inside functools.total_ordering decorator so that instnaces (Windows version data) can be compared easily.

josephsl committed:
WinVersion: introduce winVersion.getWinVer function. Re #11837.

Introduce winVersion.getWinVer function to return a WinVersion object representing the version info for the curretn system (version major, minor, build, service pack). This can be used to check for specific Windows releases (including Windows 10 feature updates) and eases comparisons among version info objects.

josephsl committed:
WinVersion: introduce winVersion.getWinVerFromVersionText function. Re #11837.

winversion.getWinVerFromVersionText function can be used to construct WinVersion class instance from the Windows version string of the form major.minor.build i.e. '10.0.19041' will return a WinVersion class instance with major=10, minor=0, build=19041. This is the text counterpart to getWinVer function and is meant to prepare for a future scenario where an add-on or other NVDA components can perform actions based on a specific version text given.

josephsl committed:
Merge branch 'master' into WinVersionClass

josephsl committed:
winVersion.WinVersion: add unit test cases. Re #11837.

Add unit tests for the following:
* winVersion.getWinVer (compared against what Python returns through sys.getwindowsversion)
* Convert a string representation of a Windows release to WinVersion class instance
* See if a specific major.minor.build is returned for a given release

josephsl committed:
winVersion unittests: lint and additional test on operators (specifically, greater than or equal i.e. newer Windows version). Re #11837

josephsl committed:
winVersion: convert Windows 10 releases to build map into str:int dictionary. Re #11795.

Convert Windows 10 release dictionary from an int:int map into an str:int map, necessary due to Version 20H2 (official documentation from Microsoft refers to October 2020 Update as 20H2, not 2009 as repoted by Windows Registry). This change is also necessary to tweak winVersion.isWin10 function (deprecated, however) to support integer arguments.

josephsl committed:
winVersion.WinVersion: transfer isWin10 function to WinVersion class. Re #11837.

Transfer the mechanism behind winversion.isWin10 function as a WinVersion class method. The new method takes string argument for release to properly recognize 20H2 (October 2020 Update). In order to guarantee compatibility for up to a year, '2009' will be transformed into '20H2' inside the new class method. Also, raise an exception instead of returning False if a given release is not a publicly released version of Windows 10.

josephsl committed:
winVersion.isWin10: deprecated, call winVersion.getWinVer().isWin10 instead. Re #11837.

Deprecate winVersion.isWin10, replaced by calling getWinVer().isWin10 function. Because the latter takes strings, convert version integers into strings before calling the WinVersion class method.

josephsl committed:
winVersion unittest: test winVersion.WinVersion.isWin10. Re #11837.

Test two aspects of isWin10 class method:
* Testing to see if running on Windows 10. On Windows 7 and 8.x, it will say 'False' by calling assertFalse.
* Test if the method raises ValueError if a nonexistent version such as '2003' is passed in.

josephsl committed:
winVersion unittest: lint

josephsl committed:
winVersion unittest: fix spelling error in comment

feerrenrut commented on tests/unit/test_winVersion.py:
```diff
@@ -0,0 +1,69 @@
+			minor=1,
+			build=7601
+		)
+		self.assertGreaterEqual(
+			winVersion.getWinVer(), minimumWinVer
+		)
```
Tests shouldn't be tied to data specific the machine they are running on, replace `winVersion.getWinVer` with a hardcoded version.

josephsl committed:
winVersion unittests/operator check: test with 'Windows 8' instead of getWinVer when checking operators. Re #11837.

An important reminder from Reef Turner (NV Access): make unit tests system-independent i.e. do not use winVersion.getWinVer function when testing operators. Instead, test 'Windows 8' a.k.a. first version to support audio ducking.

leonardder commented on source/winVersion.py:
```diff
@@ -28,38 +126,28 @@ def isUwpOcrAvailable():
 
 def isWin10(version: int = 1507, atLeast: bool = True):
 	"""
+	@deprecated: use getWinVer().isWin10 instead.
```
If we agree about this approach and I think it makes sense, we can as well remove this entirely as part of the 2021.1 release.

josephsl commented:
Hi, that could be a follow-up PR. Not only this must be done from winVersion module, other modules and add-ons must be told not to use the deprecated function. I’m happy to look into doing just that. CC @CodeOfDusk if he has any comments on this.

feerrenrut commented on source/winVersion.py:
```diff
@@ -28,38 +126,28 @@ def isUwpOcrAvailable():
 
 
 WIN10_VERSIONS_TO_BUILDS = {
```
As I understand these strings are actually 'release names'. If others agree, it would be good if we could rename this.
```suggestion
WIN10_RELEASE_NAME_TO_BUILDS = {
```

feerrenrut reviewed (changes requested):
This is looking promising, thanks @josephsl 

Can the `winVersion` variable (line 11) be removed (made private or be made into a `WinVersion` instance), or at least deprecated? I think the similar name and purpose introduces unnecessary confusion.

As part of this it would be good to be clear on the main uses cases for developers. I think generally it is easier for a developer to read / understand code that uses the release name (eg "20H2") than it is to work with the version 3-tuple. But, it is necessary to be able to convert in both directions. The design of this should encourage the in code use of the release name, while under the hood the version tuple is used.

I think one thing that is missing here is pretty printing the `WinVersion` class for the log. This should include both the tuple and the release string.

josephsl committed:
Merge branch 'master' into WinVersionClass

josephsl commented on source/winVersion.py:
```diff
@@ -15,6 +16,103 @@
+			build: int = 0,
+			servicePack: str = ""
+	):
+		if release in (None, ""):
+			self.major = major
+			self.minor = minor
+			self.build = build
+		elif release == "7":
+			self.major = 6
+			self.minor = 1
+			self.build = 7601
+		elif release == "8":
+			self.major = 6
+			self.minor = 2
+			self.build = 9200
+		elif release == "8.1":
+			self.major = 6
+			self.minor = 3
+			self.build = 9600
+		elif release in ("10", "1507"):
+			self.major = 10
+			self.minor = 0
+			self.build = 10240
+		elif release in WIN10_VERSIONS_TO_BUILDS:
+			self.major = 10
+			self.minor = 0
+			self.build = WIN10_VERSIONS_TO_BUILDS[release]
+		else:
+			raise ValueError("Cannot create Windows version information for the specified release")
+		self.servicePack = "1" if release == "7" else ""
```
Hi,

I agree with the latter point - using a pool of Windows versions (constants) not only saves us from defining a function to handle complex nature of version checks, it allows easy maintenance in the future in that it will let future contributors define a WinVersion instance representing a specific release. It also makes any part of nVDA calling winVersion module easier to read (app module handler is a key example where we check for Windows 7 versus 8 in some places and these use platform version tuple).

Thanks.

josephsl replied:
Hi,

Also, I'll go ahead and add WinVersion instances representing Windows 7 (SP1) and 8.x for sake of completeness.

Thanks.

josephsl commented on source/winVersion.py:
```diff
@@ -15,6 +16,103 @@
+			>= (other.major, other.minor, other.build)
+		)
+
+	def isWin10(self, release: str = "1507", atLeast: bool = True):
```
Hi,

Agreed, and that affirms Leonard's suggestion (removing isWin10 completely). I'll take care of this next week.

Thanks.

josephsl committed:
winVersion: add module header/description.

josephsl committed:
winVersion: record Windows releases to WinVersion instance mappings. Re #11837.

Define constants representing Windows releases to WinVersion instances (i.e. WIN7_SP1 = WinVersion(major,6, minor=1, build=7601, servicePack='1')). This eases comparisons from other NVDA modules and add-ons, along with removing the ned for a dedicated Windows 10 feature update checks. Note that server releases are not specified to avoid complicating release mappings, especially more so for Windows Server 2016 and later which are based on specific Windows 10 releases with server components added. Also, Windows 7 original release (build 7600) was added for sake of completeness.

josephsl committed:
winVersion.getWinVer: call sys.getwindowsversion function inside getWinVer, removing the need to look up old winVersion tuple. Re #11837.

Rather than using now old winVersion tuple, call sys.getwindowsversion function directly from getWinVer function.

josephsl committed:
winVersion: compare getWinVer with release constants when checking for supported OS and full screen magnification.

Testing getWinVer function: compare getWinVer output with that of Windows release constants when checking for supported OS (Windows 7 SP1) and full screen magnification (Windows 8).

josephsl committed:
Replace winVersion.isWin10 function call with constant checks. Re #11837.

Rather than using winVersion.isWin10, check Windows 10 release constants (e.g. sinWversion.WIN10_1607) directly, as these constants will record major, minor, build. Specifically:
* UIA handler and OneCore synthesizer: check winVersion.WIN10
* Windows Console support: check winVersion.WIN10_1607
* Emoji panel in May 2019 Update: check winVersion.WIN10_1903

josephsl committed:
winVersion: remove isWin10 function and test suite for it. Re #11837.

Remove winVersion.isWin10 function, both from the module and from WinVersion class, as it is now possible to compare releases via constants. Also, remove unit tests associated with this function.

josephsl committed:
winVersion.WinVersion class: simplify class constructor. Re #11837.

Remove 'release' from constructor argument and add 'product type' to mirror old winVersion.winVersion tuple data. The bulk of the old constructor (namely definig major.minor.build for a specific release) will be transferred to a new static method (next commit).

josephsl committed:
winVersion: rename Windows 10 version map to 'release name to builds' map. Re #11837.

Suggested by Reef Turner (NV Access): clarify that Windows 10 versions to builds map clearly records release names.

josephsl committed:
winVersion.WinVersion class: add two static methods to return WinVersion instance based on version string and release name. Re #11837.

Add two static methods to WinVersion class:
* fromVersionText: transfer the module-level function to this static method.
* fromReleaseName: return an instance of WinVersion class based on release (publicly documented) name.

josephsl committed:
winVersion.WinVersion: add repr method. Re #11837.

Add repr method to be used to print Windows version information to the console and to announce it at startup (future commit). It uses a helper function to obtain release name, then adds version string, service pack (if any), and product type. In essence, the work of old winVersionText was transfered to repr.

josephsl committed:
winVersion unit tests: test WinVersion.fromReleaseName functions.

Test to make sure WinVersion.fromReleaseName raises ValueError for a nonexistent release.

josephsl committed:
Remove winVersion.winVersion call. Re #11933.

Recommended by Reef Turner (Nv Access): remove winVersion.winVersion/winVersionText to reduce confusion. In preparation for it, remove winVersion.winVersion/winVersionText calls from other modules, notably app modules, app module handler, UIA objects and places where winVersion.winVersion is used. Most are replaced by winVersion.getWinVer() function call, and in case build and other information must be retrieved, whatever winVersion.getWinVer returns will be used as a starting point.

josephsl committed:
Update check: transfer winVersion.winversionText construction routine to update check module.

To guarantee backward compatiblity for OS data in update checks, transfer winVersion.winVersionText construction routine to update check module.

josephsl committed:
winVersion.WinVersion: put version major.minor.build inside parentehses in repr method.

josephsl committed:
Startup: report winVersion.getWinVer function output for Windows version. Re #11837.

Old: Using Windows version major.minor.build productType
New: Windows version: releaseName (major.minor.build) productType
Include service pack version between major.minor.build and product type if a service pack is installed (Windows 7/Server 2008 R2 only).

josephsl committed:
winVersion: remove winVersion.winVersion/winVersionText. Re #11837.

With the introduction of winVersion.WinVersion class, there is no need to keep winVersion.winVersion/winVersionText around. Therefore remove these from NVDA.

josephsl committed:
winVersion.WinVersion: say 'Windows 10 prerelease' for builds not included in public relases.

To support Windows Insider preview builds, say 'Windows 10 prerelease' if a build is not found in public releases.

josephsl committed:
winVersion.WinVersion: do not include product type string if not defined.

If WinVersion instance is built from a version text, product type will be undefined. Therefore detect this and do not include product type in output.

josephsl commented:
Hi,

Hopefully all review comments are addressed, and I have updated Add-on Updater and Windows 10 App Essentials add-ons to prepare for this PR. In the end, winVersion.winVersion/winVersionText are gone, with code paths modified to consult winVersion.getWinVer or release constants.

Thanks.

josephsl committed:
Merge branch 'master' into WinVersionClass

josephsl committed:
Update copyright year

josephsl committed:
winVersion/Windows release: shorten unknown Windows release error message.

josephsl committed:
Merged master, temporarily use master version of NVDAObjects/IAccessible/winConsole module due to conflicts

josephsl committed:
winVersion: general lint (E302, E305)

josephsl committed:
winVersion: remove unused winUser module import

josephsl committed:
NVDAObjects/IAccessible/winConsole: restore getWinVer/WIN10_1607 after resolving conflicts

lukaszgo1 commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -221,7 +227,7 @@ def event_nameChange(self, obj, nextHandler):
 		# The word "blank" is kept announced, so suppress this on build 17666 and later.
-		if winVersion.winVersion.build > 17134:
+		if winVersion.getWinVer().build > 17134:
 			# In build 17672 and later, return immediatley when element selected event on clipboard item was fired just prior to this.
```
immediatley -> immediately While it is not a mistake from this PR it makes sense to fix this while at it.

lukaszgo1 commented on source/easeOfAccess.py:
```diff
@@ -10,12 +10,12 @@
 
 # Windows >= Vista
-isSupported = winVersion.major >= 6
+isSupported = winVersion.getWinVer().major >= 6
```
Since we no longer support anything older than Win 7 this always evaluates to True. Have you considered removing this variable alltogether

lukaszgo1 commented on source/winVersion.py:
```diff
@@ -1,65 +1,186 @@
+	return getWinVer() >= WIN7_SP1
+
 
 def canRunVc2010Builds():
```
I believe this function is not used anywhere and seems to be no longer relevant. Perhaps it can be removed as part of a general clean-up of the `winVersion` module.

josephsl committed:
Merge branch 'master' into WinVersionClass

josephsl commented:
Hi, regarding the second item, I once considered removing it altogether, but figured it would be best to remove it when NVDA stops supporting Windows 7 SP1. The VC function will be gone if no-one from NVDA source code uses it. Thanks.

josephsl committed:
Modern keyboard: typo.

josephsl committed:
WinVersion: remove 'canRunVc2010Builds' function as it is no longer used by anything other than WinVersion module.

josephsl committed:
Modern keyboard: lint and update copyright header

josephsl committed:
Merge branch 'master' into WinVersionClass

feerrenrut reviewed (changes requested):
Given that `fromVersionText` and `fromReleaseName` aren't used in NVDA I think they should be removed. If people wish to make comparisons with other versions, the WinVersion class can be constructed manually.

josephsl commented:
Hi,

I'll take care of Ease of Access in a follow-up PR.

Thanks.

josephsl committed:
Merge branch 'master' into WinVersionClass

josephsl committed:
Edge Sp[artan (classic Edge): pre-gap removal = build < 15048. Re #11795

josephsl committed:
Modern keyboard/emoji panel: define variables to hold classi emoji panel releases (1709 and 1803) instead of using builds directly. Re #11795.

Comment from Reef Turner (NV Access): rather than using getWinVer().build, define variables to hold classic emoji panel releases (1709 = initial release, 1803 = feature update with proper window open event) and test membership.

josephsl committed:
WinVersion: remove fromReleaseName and fromVersionText static methods. Re #11795.

As fromVersionText and fromReleaseName are not used anywhere, remove these static methods.

josephsl committed:
WinVersion tests: remove tests that relied on removed static methods, rework nonExistentVersion test (assignment test). Re #11795.

josephsl committed:
Merge branch 'master' into WinVersionClass

josephsl commented:
While concerning, system test failure is expected. Thanks.

josephsl committed:
Merge branch 'master' into WinVersionClass

feerrenrut commented on source/appModules/windowsinternal_composableshell_experiences_textinput_inputapp.py:
```diff
@@ -173,10 +173,12 @@ def event_UIA_window_windowOpen(self, obj, nextHandler):
+		emojiPanelWindowOpenEvent = winVersion.WIN10_1803
 		if (
-			winVersion.getWinVer().build <= 17134
+			winVersion.getWinVer() in (emojiPanelInitial, emojiPanelWindowOpenEvent)
```
I may be misunderstanding (I'm not that familiar with the timeline of the emoji panel), but wouldn't this be more accurate as a comparison for a range of values eg `emojiPanelInitial  <= winVersion.getWinVer() <= emojiPanelWindowOpenEvent`. Or is it actually accurate that only for these two precise builds this alternative behavior was required.

I think generally it would be safer to do these comparisons as ranges with limits where MS made / removed things.

I ask because there are other "release" builds between, but also in case there could have been (or even could still be) people running insider builds that may still benefit from this that don't exactly equal any of the reference build numbers listed in the constants. For instance running 17763: https://blogs.windows.com/windows-insider/2018/09/18/announcing-windows-10-insider-preview-build-17763/

feerrenrut reviewed (changes requested):
Overall this looks very close to ready!

josephsl commented:
Hi, preview builds have an expiration date, so it is safe to use build lists for emoji panel as these workarounds are applicable to 1709 and 1803 only (by the way, both are out of support for consumers, and 1803 is retiring this May). The biggest change to emoji panel support won’t be done until 2022 as Python 3.10’s pattern matching syntax is perfectly applicable to emoji panel interface detection in 1809 and later (UIA Automation Id strings can be compared). Thanks.

josephsl committed:
Merge branch 'master' into WinVersionClass

josephsl committed:
Modern keyboard and unit tests: update copyright header

josephsl committed:
Modern keyboard/emoji panel: use comparisons rather than builds list.

Comment from Reef Turner (NV Access): use comparisons when checking builds. Although Windows Insider Preview builds come with expiration dates, it might be possible that not all Insiders are running latest build.

josephsl committed:
Unit tests: add comment on Flake8 F841.

feerrenrut reviewed (approved):
Thanks @josephsl

feerrenrut committed:
Update changes file for PR #11909
